src/Analyzer/Passes/UniqToCountPass.cpp

```diff
@@ -0,0 +1,142 @@
+{
+public:
+    using Base = InDepthQueryTreeVisitorWithContext<UniqToCountVisitor>;
+    using Base::Base;
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        if (!getSettings().optimize_uniq_to_count)
+            return;
+
+        auto * query_node = node->as<QueryNode>();
+        if (!query_node)
+            return;
+
+        /// Check that query has only single table expression which is subquery
+        auto * subquery_node = query_node->getJoinTree()->as<QueryNode>();
+        if (!subquery_node)
+            return;
+
+        /// Check that query has only single node in projection
+        auto & projection_nodes = query_node->getProjection().getNodes();
+        if (projection_nodes.size() != 1)
+            return;
+
+        /// Check that projection_node is a function
+        auto & projection_node = projection_nodes[0];
+        auto * function_node = projection_node->as<FunctionNode>();
+        if (!function_node)
+            return;
+
+        /// Check that query single projection node is `uniq` or its variants
+        if (!matchFnUniq(function_node->getFunctionName()))
+            return;
+
+        auto & uniq_arguments_nodes = function_node->getArguments().getNodes();
+
+        /// Whether query matches 'SELECT uniq(x ...) FROM (SELECT DISTINCT x ...)'
+        auto match_subquery_with_distinct = [&]() -> bool
+        {
+            if (!subquery_node->isDistinct())
+                return false;
+            /// uniq expression list == subquery group by expression list
+            if (!nodeListEquals(uniq_arguments_nodes, subquery_node->getProjection().getNodes()))
+                return false;
+            return true;
+        };
+
+        /// Whether query matches 'SELECT uniq(x ...) FROM (SELECT x ... GROUP BY x ...)'
+        auto match_subquery_with_group_by = [&]() -> bool
+        {
+            if (!subquery_node->hasGroupBy())
+                return false;
+            /// uniq argument node list == subquery group by node list
+            if (!nodeListEquals(uniq_arguments_nodes, subquery_node->getGroupByNode()->getChildren()))
+                return false;
+            /// subquery select node list must contain all columns in uniq argument node list
+            if (!nodeListContainsAll(subquery_node->getProjection().getNodes(), uniq_arguments_nodes))
+                return false;
+            return true;
+        };
+
+        /// Replace uniq of initial query to count
+        if (match_subquery_with_distinct() || match_subquery_with_group_by())
+        {
+            AggregateFunctionProperties properties;
+            auto aggregate_function = AggregateFunctionFactory::instance().get("count", {}, {}, properties);
+            function_node->resolveAsAggregateFunction(std::move(aggregate_function));
+            function_node->getArguments().getNodes().clear();
+            query_node->resolveProjectionColumns({{"count()", function_node->getResultType()}});
```

Comment:
Comment in `QueryAnalysisPass` shows we should not modify projection columns later. But if not, result of `explain query tree` will have inconsistent result. I am not sure how to handle it, please give a hand.

Reply:
@novikd you mean just ignore `resolveProjectionColumns`?